lib/navigation/preprocess_encoders.cpp

```diff
@@ -16,41 +31,89 @@ core::EncoderData EncodersPreprocessor::processData(const core::EncoderData)
   - call checkRelaible
 
   */
-  return {0, 0, 0, 0};
+  SensorChecks sensors  = checkReliable();
+  if(sensors == SensorChecks::kUnacceptable){
+    return std::nullopt;
+  }
+ return detectOutliers(encoder_data);
 }
 
+template<std::size_t N>
+core::Float EncodersPreprocessor::getQuartile(const std::array<std::uint32_t, N> &reliable_data, const core::Float quartile_percent){
+
+  const core::Float quartile_index = (num_reliable_encoders_ - 1)*quartile_percent;
+  const std::uint8_t quartile_high = static_cast<int>(std::ceil(quartile_index));
+  const std::uint8_t quartile_low = static_cast<int>(std::floor(quartile_index));
+  const core::Float quartile = (reliable_data.at(quartile_high) + reliable_data.at(quartile_low))/2.0;
+  return quartile;
+}
+
+template<std::size_t N>
+Quartile EncodersPreprocessor::getQuartiles(std::array<std::uint32_t, N> &reliable_data){
+  std::sort(reliable_data.begin(), reliable_data.end());
+  Quartile quartile;
+  quartile.q1 = getQuartile(reliable_data,0.25);
+  quartile.median = getQuartile(reliable_data, 0.5);
+  quartile.q3 = getQuartile(reliable_data, 0.75);
+  return quartile;
+}
+
+
 core::EncoderData EncodersPreprocessor::detectOutliers(const core::EncoderData encoder_data)
 {
-  /*
-  TODOLater: implement
-  rough process:
-  - get q1, median, q3 of encoder array
-  - define upper & lower bounds as (-)1.5*inter-quatrile range
-  - if any datapoint outwith this, set
-  num_outliers_per_encoder_[i] += 1
-  -set outlier points as median
-
-  -also has to be able to handle 1 unreliable sensor
-  -also figure out return type/ what we update and update
-  documentation as appropriate
-  */
-  return {0, 0, 0, 0};
+  core::EncoderData encoder_data_copy;
+  std::copy(encoder_data.begin(), encoder_data.end(), encoder_data_copy.begin());
+  Quartile quartiles;
+  if (num_reliable_encoders_ == core::kNumEncoders - 1) {
+    std::array<uint32_t, core::kNumEncoders - 1> reliable_data;
+    std::size_t counter = 0;
+    for (std::size_t i = 0; i < encoder_data.size(); ++i) {
+      if (encoders_reliable.at(i)) {
+        reliable_data.at(counter) = encoder_data.at(i);
+        ++counter;
+      }
+    }
+    quartiles = getQuartiles(reliable_data);
+  } else {
+    core::EncoderData reliable_data;
+    std::copy(
+      encoder_data.begin(), encoder_data.end(), reliable_data.begin());
+    quartiles = getQuartiles(reliable_data); 
+  }
+  const core::Float iqr         = quartiles.q3 - quartiles.q1;
```

Comment:
I know its pedantic but maybe `interquartile_range` or a comment explaining what it is for someone reading this without context.